README
User configuration notes
========================

Below is a quick guide to necessary configurations for getting
started with the Erlang mode for Emacs. Please refer to the
Users guide and reference manual in the documentation for the
Erlang/OTP application tools for more information.

You can obtain a copy of the Erlang man pages here:

    http://www.erlang.org/download

Note for the man page menus to be populated, the Erlang
man pages must be in uncompressed form.

Additional documenation for this mode can be found here:

    http://erlware.org/tools/erlware-mode


For UNIX users
--------------

To set up the Erlang Emacs mode on UNIX systems, edit/create the file
.emacs in the your home directory.

Below is a complete example of what should be added to a user's .emacs
provided that OTP is installed in the directory /usr/local/otp:
        
      (setq load-path (cons "<path-to-mode>/erlware-mode" load-path))
      (setq erlang-man-root-dir "/usr/local/otp")
      (setq exec-path (cons "/usr/local/otp/bin" exec-path))
      (require 'erlang-start)

For XEmacs, you use add-to-list to modify the load path:

      (add-to-list 'load-path "<path-to-mode>/erlware-mode")

If your XEmacs system already has the default erlang mode on
the search path, you may need to put the erlware path at the
front of the search path:

      (push "<path-to-mode>/erlware-mode" load-path)


For Windows users
-----------------

To set up the Erlang Emacs mode on Windows systems, edit/create the
file .emacs, the location of the file depends on the configuration of
the system. If the HOME environment variable is set, Emacs will look
for the .emacs file in the directory indicated by the HOME variable.
 If HOME is not set, Emacs will look for the .emacs file in C:\.

Below is a complete example of what should be added to a user's .emacs
provided that OTP is installed in the directory C:\Program
Files\erl-<Ver>:

      (setq load-path (cons  "<path-to-mode>/erlware-mode" load-path))
      (setq erlang-man-root-dir "C:/Program Files/erl<Ver>")
      (setq exec-path (cons "C:/Program Files/erl<Ver>/bin" exec-path))
      (require 'erlang-start)

For XEmacs, you use add-to-list to modify the load path:

      (add-to-list 'load-path "<path-to-mode>/erlware-mode")

If your XEmacs system already has the default erlang mode on
the search path, you may need to put the erlware path at the
front of the search path:

      (push "<path-to-mode>/erlware-mode" load-path)


ChangeLog
2009-12-28  Dave Peticolas <dave@krondo.com>

	* Add -define(SERVER, ?MODULE). to gen_event to fix
	compilation. Add newline.

2009-12-21  bile <bile@landofbile.com>

	* Add -define(SERVER, ?MODULE). to fix compilation

	* Added underscores to handle_sync_event's Event and From to
	suppress warning.

2009-12-21  bile <bile@landofbile.com>

	* Create new option, erlang-compile-outdir, which is concat'ed to
	the beam outdir when calling erlang-compile

2009-11-09  Tino Breddin <tino@.(none)>

	* changed shebang to use env instead of hardcoded path

2009-09-14  Magnus Henoch <magnus@erlang-consulting.com>

	* erlang.el (erlang-dollar-is-escape): New variable.
        (erlang-toggle-dollar-is-escape): New function.
        (erlang-syntax-table-init): Check erlang-dollar-is-escape.

2009-07-28  Magnus Henoch <magnus.henoch@gmail.com>

	* erlang-start.el: Associate files using interpreter "escript" with Erlang mode.

2009-07-16  mats cronqvist <mats.cronqvist@gmail.com>

	* better include file detection (again)

2009-07-15  mats cronqvist <mats.cronqvist@gmail.com>

	* removed flawed next-error code.

2009-06-15  mats cronqvist <mats.cronqvist@gmail.com>

	* this should have been part of the previous commit.

2009-06-15  mats cronqvist <mats.cronqvist@gmail.com>

	* make flymake use $HOME/.erlang-flymake as temp dir. Ought to work on windows, and (more importantly) if the erl file is write protected.

2009-06-02  mats cronqvist <mats.cronqvist@gmail.com>

	* fixed the backtick problems by removing emacs 19 compatibility flymake enhancements

2009-07-02  Magnus Henoch <magnus@erlang-consulting.com>

	* erlang.el (erlang-calculate-stack-indent): Combine ignored 'if' statement for 'after' with the one for 'catch', to fix indentation of try-after.

2009-06-27  Dave Peticolas <dave@krondo.com>

	* Ignore module name mismatch, as flymake renames the file.

2009-06-11  Magnus Henoch <magnus@erlang-consulting.com>

	* erlang.el (erlang-get-function-name-and-arity): Fix arity: only count "sexp" as parameter when followed by "," or ")".

2009-06-08  Magnus Henoch <magnus@erlang-consulting.com>

	* erlang.el (erlang-get-function-name-and-arity): New function. (erlang-mode-variables): Use it to include function arity in imenu entries.

2009-05-22  Magnus Henoch <magnus@erlang-consulting.com>

	* (erlang-mode-variables): Set comment-add to 0 if using new-style comments, for proper operation of M-; on regions.

2009-05-18  Magnus Henoch <magnus@erlang-consulting.com>

	* (erlang-mode-map): Define keymap when module is loaded.
 	
	* (erlang-mode): Don't call erlang-keymap-init.

	* (erlang-keymap-init, erlang-mode-commands): Remove.

2009-05-08  Magnus Henoch <magnus@erlang-consulting.com>

	* (inferior-erlang): Prompt for command when called with prefix argument.

2009-05-14  mats cronqvist <mats.cronqvist@gmail.com>

	* flymake support

2009-05-04  Dave Peticolas <dave@krondo.com>

	* Jim Morris's patch to make erlang.el byte-compile in 22.1 and 22.2.

2009-05-04  Dave Peticolas <dave@krondo.com>

	* Support tags in Emacs > 19. (ported from original mode by Jim Morris).

2009-03-20  mats cronqvist <mats.cronqvist@gmail.com>

	* numerous syntax highlighting fixes

2009-02-19  Derek Upham <sand@blarg.net>

	* run after-change-major-mode-hook in Emacs 23 and greater

2009-02-18  mats cronqvist <mats.cronqvist@gmail.com>

	* himBHliteral fix

2009-02-16  mats cronqvist <mats.cronqvist@gmail.com>

	* fixed syntax hilite

2008-12-11  Dave Peticolas <dave@krondo.com>

	* Fix bif/guard highlighting (Issue 25).

2008-10-31  Dave Peticolas <dave@krondo.com>

	* Remove cruft. Fix issue 24 (catch clause indentation).

2008-10-31  Dave Peticolas <dave@krondo.com>

	* Update tests.

2008-05-17  Dave Peticolas <dave@krondo.com>

	* Remove the arrow font lock function.

2008-05-16  Dave Peticolas <dave@krondo.com>

	* Erlang test file for erlware mode.

2008-05-15  Dave Peticolas <dave@krondo.com>

	* Remove dead code.

2008-05-15  Dave Peticolas <dave@krondo.com>

	* Treat $ as a character quote, not an expression prefix.

2008-05-04  Dave Peticolas <dave@krondo.com>

	* Automatic newline for electric comma at the end of erlang clauses.

2008-04-28  Dave Peticolas <dave@krondo.com>

	* New comment indenting behavior.

2008-04-20  Dave Peticolas <dave@krondo.com>

	* Use supervisor:start_link/0. Update docs.

2008-04-12  Dave Peticolas <dave@krondo.com>

	* Consistent spacing in export lines.

2008-04-04  Dave Peticolas <dave@krondo.com>

	* Remove unused files. The app-wizard is superseded by sinan.

2008-02-03  Dave Peticolas <dave@krondo.com>

	* Use newline-and-indent.

2008-02-03  Dave Peticolas <dave@krondo.com>

	* Turn on electric newline by default.

2008-02-03  Dave Peticolas <dave@krondo.com>

	* Apply separator config to section headers.

2008-02-03  Dave Peticolas <dave@krondo.com>

	* Consistently use 3 percent signs for section headers.

2008-02-03  Dave Peticolas <dave@krondo.com>

	* Consistent newline after section headers.

2008-02-03  Dave Peticolas <dave@krondo.com>

	* Configurable horizontal separators for function docs.

2008-01-09  Dave Peticolas <dave@krondo.com>

	* Add erlware.org to README.

2008-01-05  Dave Peticolas <dave@krondo.com>

	* Update README with erlare-man-root-dir change.

2008-01-05  Dave Peticolas <dave@krondo.com>

	* Rename erlang-root-dir to erlang-man-root-dir.

2008-01-03  Dave Peticolas <dave@krondo.com>

	* Add sinan clean function.

2008-01-01  Dave Peticolas <dave@krondo.com>

	* Start of sinan command support.

2007-12-29  Dave Peticolas <dave@krondo.com>

	* Obey user's delete-key-deletes-forward setting.

2007-12-27  Dave Peticolas <dave@krondo.com>

	* Update supervisor init method.

2007-12-27  Dave Peticolas <dave@krondo.com>

	* Distinguish between original mode and erlware mode.

2007-12-27  Dave Peticolas <dave@krondo.com>

	* Whitespace cleanup.

2007-12-26  Martin Logan <dave@krondo.com>

	* updated the init method


erlang-start.el
;; erlang-start.el --- Load this file to initialize the Erlang package.

;; Copyright (C) 1998  Ericsson Telecom AB

;; Author:   Anders Lindgren
;; Version:  2.3
;; Keywords: erlang, languages, processes
;; Created:  1996-09-18
;; Date:     1998-03-16

;;; Commentary:

;; Introduction:
;; ------------
;;
;; This package provides support for the programming language Erlang.
;; The package provides an editing mode with lots of bells and
;; whistles, compilation support, and it makes it possible for the
;; user to start Erlang shells that run inside Emacs.
;;
;; See the Erlang distribution for full documentation of this package.

;; Installation:
;; ------------
;;
;; Place this file in Emacs load path, byte-compile it, and add the
;; following line to the appropriate init file:
;;
;;    (require 'erlang-start)
;;
;; The full documentation contains much more extensive description of
;; the installation procedure.

;; Reporting Bugs:
;; --------------
;;
;; Please send bug reports to the following email address:
;;     support@erlang.ericsson.se
;;
;; Please state as exactly as possible:
;;    - Version number of Erlang Mode (see the menu), Emacs, Erlang,
;;	and of any other relevant software.
;;    - What the expected result was.
;;    - What you did, preferably in a repeatable step-by-step form.
;;    - A description of the unexpected result.
;;    - Relevant pieces of Erlang code causing the problem.
;;    - Personal Emacs customisations, if any.
;;
;; Should the Emacs generate an error, please set the emacs variable
;; `debug-on-error' to `t'.  Repeat the error and enclose the debug
;; information in your bug-report.
;;
;; To set the variable you can use the following command:
;;     M-x set-variable RET debug-on-error RET t RET

;;; Code:

;;
;; Declare functions in "erlang.el".
;;

(autoload 'erlang-mode "erlang" "Major mode for editing Erlang code." t)
(autoload 'erlang-version "erlang"
  "Return the current version of Erlang mode." t)
(autoload 'erlang-shell "erlang" "Start a new Erlang shell." t)
(autoload 'run-erlang "erlang" "Start a new Erlang shell." t)

(autoload 'erlang-compile "erlang"
  "Compile Erlang module in current buffer." t)

(autoload 'erlang-man-module "erlang"
  "Find manual page for MODULE." t)
(autoload 'erlang-man-function "erlang"
  "Find manual page for NAME, where NAME is module:function." t)

(autoload 'erlang-find-tag "erlang"
  "Like `find-tag'.  Capable of retreiving Erlang modules.")
(autoload 'erlang-find-tag-other-window "erlang"
  "Like `find-tag-other-window'.  Capable of retreiving Erlang modules.")


;;
;; Associate files extensions ".erl" and ".hrl" with Erlang mode.
;;

(let ((a '("\\.erl\\'" . erlang-mode))
      (b '("\\.hrl\\'" . erlang-mode)))
  (or (assoc (car a) auto-mode-alist)
      (setq auto-mode-alist (cons a auto-mode-alist)))
  (or (assoc (car b) auto-mode-alist)
      (setq auto-mode-alist (cons b auto-mode-alist))))

;;
;; Associate files using interpreter "escript" with Erlang mode.
;; 

(add-to-list 'interpreter-mode-alist (cons "escript" 'erlang-mode))

;;
;; Ignore files ending in ".jam", ".vee", and ".beam" when performing
;; file completion.
;;

(let ((erl-ext '(".jam" ".vee" ".beam")))
  (while erl-ext
    (let ((cie completion-ignored-extensions))
      (while (and cie (not (string-equal (car cie) (car erl-ext))))
	(setq cie (cdr cie)))
      (if (null cie)
	  (setq completion-ignored-extensions
		(cons (car erl-ext) completion-ignored-extensions))))
    (setq erl-ext (cdr erl-ext))))


;;
;; The end.
;;

(provide 'erlang-start)

;; erlang-start.el ends here.
